subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java

```diff
@@ -61,26 +61,45 @@ private PersistedData serializeEntry(Map.Entry<K, V> entry, PersistedDataSeriali
 
     @Override
     public Optional<Map<K, V>> deserialize(PersistedData data) {
-        if (!data.isArray()) {
+        if (!data.isArray() || data.isValueMap()) {
+            logger.warn("Incorrect map format detected: object instead of array.\n" +
+                    "Expected\n" +
+                    "  \"mapName\": [\n" +
+                    "    { \"key\": \"...\", \"value\": \"...\" }\n" +
+                    "  ]\n" +
+                    "but found \n'{}'", data);
             return Optional.empty();
         }
 
         Map<K, V> result = Maps.newLinkedHashMap();
 
         for (PersistedData entry : data.getAsArray()) {
             PersistedDataMap kvEntry = entry.getAsValueMap();
-            final Optional<K> key = keyHandler.deserialize(kvEntry.get(KEY));
-
-            if (key.isPresent()) {
-                final Optional<V> value = valueHandler.deserialize(kvEntry.get(VALUE));
-                if (value.isPresent()) {
-                    result.put(key.get(), value.get());
-                } else {
-                    logger.warn("Missing value for key '{}' with {} given entry '{}'", key.get(), valueHandler, kvEntry.get(VALUE));
-                }
-            } else {
-                logger.warn("Missing field '{}' for entry '{}'", KEY, kvEntry);
+            PersistedData rawKey = kvEntry.get(KEY);
+            PersistedData rawValue = kvEntry.get(VALUE);
+            if (rawKey == null || rawValue == null) {
+                logger.warn("Incorrect map format detected: missing map entry with \"key\" or \"value\" key.\n" +
+                        "Expected\n" +
+                        "  \"mapName\": [\n" +
+                        "    { \"key\": \"...\", \"value\": \"...\" }\n" +
+                        "  ]\n" +
+                        "but found \n'{}'", data);
+                return Optional.empty();
             }
+
+            final Optional<K> key = keyHandler.deserialize(rawKey);
+            if (key.isEmpty()) {
+                logger.warn("Could not deserialize key '{}' as '{}'", rawKey, keyHandler.getClass().getSimpleName());
+                return Optional.empty();
```

Comment:
@keturn I saw the test cases in `GenericMapTypeHandlerTest.java` and wanted to ask, why it's desirable to have an empty map returned in case of a mismatched key / value handler... My current assumption is that we want that because failing to deserialize the `ChangingBlocks` component should not result in the whole prefab deserialization failing (although I'm not even sure that would be the case when returning `Optional.empty()` instead :thinking: 

Can you confirm this assumption or otherwise explain the reason for expecting an empty map in this case?

My intention with using `Optional.empty()` here is to indicate that the serialization failed and to not return a half-way deserialized map as might happen for instance in case the prefab holds:
```
myMap: [
  { "key": "myKey", "value": "myValue" },
  { "key2": "myKey2", "value2": "myValue2" }
]
```
With returning `Optional.of(result)`, I would expect the first map entry to be put into the result map and the deserialization failing for the second entry, resulting in a half-way deserialized map which might lead to weird and potentially hard to debug behavior in game.
Taking the growing plant feature, for example, the plant would never grow to the second stage as the second stage never ended up in game because its deserialization failed.

Please correct me if I got any of the type handling logic etc. wrong leading to incorrect assumptions/expectations :sweat_smile:

Reply:
It looks like I wrote those test cases after @skaldarnar wrote the implementation, so we should ask that guy too.

I'm struggling with how to answer these questions because I think you raise some good points, and—

- I'm not sure how to give advice on the best use of `Optional.empty` here, because I feel like `Optional` is bad at communicating error states and I've felt frustrated with trying to use it here myself, and
- You don't want the result to be a partly-filled data structure. I empathize, but I think a partial response like that would be more consistent with the way the other deserialization methods are written. If I understand correctly, it seems to be working with a best-effort approach, where it'll do its best to return something even if it might have holes in it.